src/sqlfluff/core/rules/base.py

```diff
@@ -798,20 +798,26 @@ def _choose_anchor_segment(context, edit_type, segment, filter_meta=False):
         anchor = segment
         child = segment
         for seg in context.parent_stack[0].path_to(segment)[1:-1][::-1]:
-            children = (
-                seg.segments
-                if not filter_meta
-                else [child for child in seg.segments if not child.is_meta]
-            )
-            if edit_type == "create_before" and children[0] is child:
-                anchor = seg
-                assert anchor.raw.startswith(segment.raw)
-            elif edit_type == "create_after" and children[-1] is child:
-                anchor = seg
-                assert anchor.raw.endswith(segment.raw)
-            else:
-                break
-            child = seg
+            # Which sets of children to check against.
+            childrenses = []
+            if filter_meta:
+                # Optionally check against filtered (non-meta only) children.
+                childrenses.append(
+                    [child for child in seg.segments if not child.is_meta]
+                )
+            # Always check against the full set of children.
+            childrenses.append(seg.segments)
```

Comment:
does this want an else??

Reply:
No, we want to check this regardless. I found a corner case where if the rule wants to anchor to a meta segment (such as a `Dedent`), we need to check against `seg.segments`, even if we were asked to ignore meta segments.

The details here can become a bit confusing. The high-level goal is to walk up the parse tree towards the top-level file segment as long as the higher-level segments end/start at the same file location as the one below. This is kind of a hack, but a useful one -- when adding some segments such as whitespace, comments, and semicolons, we want to put them higher in the stack because that's how the grammar rules seem to want them. Someday, this may need to be replaced with something smarter, but this has already been helpful for fixing post-fix parse issues with several rules, which is 😎 .

Reply:
But what I don't understand is:
if filter_meta:
  append all non-meta
  then append everything?
  
won't that mean segments are there twice?

Reply:
Look carefully -- `childrenses` is a list of lists. If it finds a match using _either_ of the lists, it proceeds up the tree. So if it finds a match using the first one, it ignores the second one.

Reply:
oh, gotcha - makes sense now 😅